keps/sig-apps/2214-indexed-job/README.md

```diff
@@ -0,0 +1,487 @@
+Jobs, this translates as more than one Pod having the same index.
+
+Just like for existing Job patterns, workloads have to handle duplicates at the
+application level.
+
+[reference]: https://kubernetes.io/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures
+
+## Design Details
+
+### Indexed Jobs
+
+Indexed Jobs are Jobs [with fixed completion count], that is, Jobs that have a
+non-nil `.spec.completions` field.
+
+### Pod detail
+
+The Pod and PodSpec APIs don't get any new fields. However, Pods created for
+Indexed Jobs get the annotation `alpha.batch.k8s.io/job-completion-index` with a
+value equal to its completion index. The annotation is immutable.
+
+The annotation can be accessed through the downward API as a file or environment
+variable.
+
+### Job completion and restart policy
+
+When dealing with Indexed Jobs, the Job controller keeps track of Pod
+completions for each index from 0 to `.spec.completions - 1`. When a Pod
+finishes successfully, the controller doesn't create another Pod with the same
+completion index. The Job controller doesn't remove succeeded or failed Pods.
+
+If an entire Pod fails, such as when the Pod gets kicked off the node or if a
+container of the Pod fails and `.spec.template.spec.restartPolicy = "Never"`,
+the Job controller gets the completion index for the failed Pod and creates a
+new Pod with the same index. The application needs to handle restarts in a
+different Pod.
+
+The kubelet handles container restarts as usual, according to the
+`spec.template.spec.restartPolicy`.
+
+### Job scaling
+
+Jobs can be scaled up and down by varying `.spec.parallelism` (note that
+`.spec.completions` is an immutable field).
+
+When starting or scaling up a Job, the Job controller creates Pods with lower
+completion index first, as long as there is no other completed or running Pod
+with the same index. This is to make the controller behavior more
+predictable. We do not offer any guarantees.
+
+Scale down is unaffected by completion index.
+
+### Test Plan
+
+Unit, integration and E2E tests cover the following Indexed Job mechanics:
+
+  - Creation with indexed Pod names and index annotations.
+  - Scale up and down.
+  - Pod failures.
+  
+### Graduation Criteria
+
+#### Alpha -> Beta Graduation
+
+- Complete features:
+  - Completion index in annotation
+  - Restart policy
+- Gather feedback from end users and operators' developers. Open questions:
+  - Are stable Pod names necessary?
+- Tests are in Testgrid and linked in KEP
+
+#### Beta -> GA Graduation
+
+- E2E test graduates to conformance
+- 2 examples of operators using Indexed Jobs.
+
+### Upgrade / Downgrade Strategy
+
+The Job controller keeps track of running or succeeded indexes using the
+annotation. In the event of a kube-controller-manager upgrade, some running or
+succeeded Pods might exist without a completion index.
+
+<<[UNRESOLVED handling Jobs with existing Pods without completion index ]>>
```

Comment:
@wojtek-t WDYT?

Reply:
From @erictune:

> Did you consider the alternative where you do change the API such as by adding .spec.indexed boolean?
>
> Some pros of that approach are:
>
> 1. Could use a different path in the control code for indexed jobs. Reduces risk that the new path in the controller hurts existing multi-completition jobs that don't need indexes.
> 1. Upgrade downgrade is simple: On downgrade, all the ones with ".spec.indexed" stop working. On upgrade, those without ".spec.indexed" don't do anything different.
> 1. Can inject index without modifying existing pods see subsequent comments).

My response:

Yes, I have considered the idea, and it's open for discussion.

why is (3) important? The index would always be added, but Pods might not use it. So, in general, Pods don't need changes.
I don't love the idea of having a field just because of upgrades (2); WDYT of my alternatives to this? (1) is a more compelling reason, but risk can be minimized with proper coverage.

Perhaps @liggitt can chime in as well

Reply:
Sorry - I think I didn't understand the problem yet. Why those pods may not have those annotations with index?

Reply:
Because they were created by an older version of the controller that didn't add them.